proposals/intraprocess_communication.md

```diff
@@ -0,0 +1,525 @@
+
+Even if ROS2 supports intra-process communication, the implementation of this mechanism has still much space for improvement.
+Until ROS2 Crystal, major performance issues and the lack of support for shared pointer messages were preventing the use of this feature in real applications.
+
+With the ROS2 Dashing release, most of these issues have been addressed and the intra-process communication behavior has improved greatly ([see ticket](https://github.com/ros2/ros2/issues/649)).
+
+The current implementation is based on the creation of a ring buffer for each `Publisher` and on the
+publication of meta-messages through the middleware layer. When a `Publisher` has to publish intra-process, it
+will pass the message to the `IntraProcessManager`. Here the message will be stored in the ring buffer
+associated with the `Publisher`. In order to extract a message from the `IntraProcessManager` two pieces of
+information are needed: the id of the `Publisher` (in order to select the correct ring buffer) and the
+position of the message within its ring buffer. A meta-message with this information is created and sent
+through the ROS2 middleware to all the `Subscription`s, which can then retrieve the original message from the
+`IntraProcessManager`.
+
+![Current IPC Block Diagram](old_ipc.png)
+
+Several shortcomings of the current implementation are listed below.
+
+### Incomplete Quality of Service support
+
+The current implementation can't be used when the QoS durability value is set to `Transient Local`.
+
+The current implementation does not enforce the depth of the QoS history in a correct way. The reason is that
+there is a single ring buffer per `Publisher` and its size is equal to the depth of the `Publisher`'s history.
+A `Publisher` stores a message in the ring buffer and then it sends a meta-message to allow a `Subscription` to retrieve it.
+The `Subscription` correctly stores meta-messages up to the number indicated by its depth of the history, but,
+depending on the frequency at which messages are published and callbacks are triggered, it may happen that a
+meta-message processed from the `Subscription` does not correspond anymore to a valid message in the ring
+buffer, because it has been already overwritten.
+
+More details [here](https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/).
+
+**TODO:** take into account also new QoS: Deadline, Liveliness and Lifespan
+[reference](https://github.com/ros2/design/pull/212).
+
+### Dependent on the RMW
+
+The current implementation has to send meta-messages from the `Publisher` to the `Subscription`s. This is done
+using the `rmw_publish` function, the implementation of which depends on the chosen middleware. This results
+in having the performance of a ROS2 application with intra-process communication enabled to still heavily
+depend on the chosen RMW implementation.
+
+At the moment, this results in a relevant impact on the performance of the application for
```

Comment:
I'm not entirely sure what you mean by this sentence.  Could you perhaps clarify?

Reply:
Rephrased as:
```
Given the fact that these meta-messages have only to be received from entities within 
the same process,  there is space for optimizing how they are transmitted by each RMW.

However, at the moment none of the supported RMW is actively tackling this issue.

This results in that the performance of a single process ROS2 application with 
intra-process communication enabled are still worst than what you could expect 
from a non-ROS application sharing memory between its components.
```